File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/HashesShouldHaveUnpredictableSalt.NetCore.cs
using System;
using System.Security.Cryptography;

public interface InterfaceWithMethodImplementation
{
    public void Method(string password)
    {
        var salt = new byte[16];
        new Rfc2898DeriveBytes(password, salt);                                                                                 // Noncompliant
    }
}

// The Rfc2898DeriveBytes.Pbkdf2 method is available in .NET 6.0+
public class Pbkdf2Method
{
    public void DeriveKeyWithPbkdf2Method(byte[] passwordBytes, string passwordText)
    {
        var rng = RandomNumberGenerator.Create();
        var unsafeSalt = new byte[15];
        var safeSalt = new byte[16];
        rng.GetBytes(safeSalt);

        var unsafeSpan = unsafeSalt.AsSpan();
        var unsafeReadonlySpan = new ReadOnlySpan<byte>(unsafeSalt);
        var safeSpan = safeSalt.AsSpan();
        var safeReadonlySpan = new ReadOnlySpan<byte>(safeSalt);

        var key = Rfc2898DeriveBytes.Pbkdf2(passwordBytes, unsafeSalt, 42, HashAlgorithmName.SHA512, 42);                       // Noncompliant
        key = Rfc2898DeriveBytes.Pbkdf2(passwordText, unsafeSalt, 42, HashAlgorithmName.SHA512, 42);                            // Noncompliant
        key = Rfc2898DeriveBytes.Pbkdf2(passwordText.AsSpan(), unsafeSalt, 42, HashAlgorithmName.SHA512, 42);                   // Noncompliant

        key = Rfc2898DeriveBytes.Pbkdf2(passwordBytes, unsafeSpan, 42, HashAlgorithmName.SHA512, 42);                           // Noncompliant
        key = Rfc2898DeriveBytes.Pbkdf2(passwordBytes, unsafeReadonlySpan, 42, HashAlgorithmName.SHA512, 42);                   // Noncompliant
        key = Rfc2898DeriveBytes.Pbkdf2(passwordBytes, unsafeSalt.AsSpan(), 42, HashAlgorithmName.SHA512, 42);                  // Noncompliant
        key = Rfc2898DeriveBytes.Pbkdf2(passwordBytes, new Span<byte>(unsafeSalt), 42, HashAlgorithmName.SHA512, 42);           // Noncompliant
        key = Rfc2898DeriveBytes.Pbkdf2(passwordBytes, new ReadOnlySpan<byte>(unsafeSalt), 42, HashAlgorithmName.SHA512, 42);   // Noncompliant
        key = Rfc2898DeriveBytes.Pbkdf2(passwordBytes.AsSpan(), unsafeSpan, 42, HashAlgorithmName.SHA512, 42);                  // Noncompliant

        key = Rfc2898DeriveBytes.Pbkdf2(passwordBytes, safeSalt, 42, HashAlgorithmName.SHA512, 42);                             // Compliant
        key = Rfc2898DeriveBytes.Pbkdf2(passwordBytes, safeSpan, 42, HashAlgorithmName.SHA512, 42);
        key = Rfc2898DeriveBytes.Pbkdf2(passwordBytes, safeReadonlySpan, 42, HashAlgorithmName.SHA512, 42);
        key = Rfc2898DeriveBytes.Pbkdf2(passwordBytes, safeSalt.AsSpan(), 42, HashAlgorithmName.SHA512, 42);
        key = Rfc2898DeriveBytes.Pbkdf2(passwordBytes, new Span<byte>(safeSalt), 42, HashAlgorithmName.SHA512, 42);
        key = Rfc2898DeriveBytes.Pbkdf2(passwordBytes, new ReadOnlySpan<byte>(safeSalt), 42, HashAlgorithmName.SHA512, 42);
        key = Rfc2898DeriveBytes.Pbkdf2(passwordBytes.AsSpan(), safeSpan, 42, HashAlgorithmName.SHA512, 42);
    }
}

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/HashesShouldHaveUnpredictableSalt.NetCore.vb
Imports System.Security.Cryptography

' The Rfc2898DeriveBytes.Pbkdf2 method is available in .NET 6.0+
Public Class Pbkdf2Method
    Public Sub DeriveKeyWithPbkdf2Method(passwordBytes As Byte(), passwordText As String)
        Dim rng = RandomNumberGenerator.Create()
        Dim unsafeSalt = New Byte(14) {}
        Dim safeSalt = New Byte(15) {}
        rng.GetBytes(safeSalt)

        Dim unsafeSpan = unsafeSalt.AsSpan()
        Dim safeSpan = safeSalt.AsSpan()

        Dim key = Rfc2898DeriveBytes.Pbkdf2(passwordBytes, unsafeSalt, 42, HashAlgorithmName.SHA512, 42)                        ' Noncompliant
        key = Rfc2898DeriveBytes.Pbkdf2(passwordText, unsafeSalt, 42, HashAlgorithmName.SHA512, 42)                             ' Noncompliant
        key = Rfc2898DeriveBytes.Pbkdf2(passwordText.AsSpan(), unsafeSalt, 42, HashAlgorithmName.SHA512, 42)                    ' Noncompliant

        key = Rfc2898DeriveBytes.Pbkdf2(passwordBytes, unsafeSpan, 42, HashAlgorithmName.SHA512, 42)                            ' Noncompliant
        key = Rfc2898DeriveBytes.Pbkdf2(passwordBytes, unsafeSalt.AsSpan(), 42, HashAlgorithmName.SHA512, 42)                   ' Noncompliant
        key = Rfc2898DeriveBytes.Pbkdf2(passwordBytes.AsSpan(), unsafeSpan, 42, HashAlgorithmName.SHA512, 42)                   ' Noncompliant

        key = Rfc2898DeriveBytes.Pbkdf2(passwordBytes, safeSalt, 42, HashAlgorithmName.SHA512, 42)                              ' Compliant
        key = Rfc2898DeriveBytes.Pbkdf2(passwordBytes, safeSpan, 42, HashAlgorithmName.SHA512, 42)
        key = Rfc2898DeriveBytes.Pbkdf2(passwordBytes, safeSalt.AsSpan(), 42, HashAlgorithmName.SHA512, 42)
        key = Rfc2898DeriveBytes.Pbkdf2(passwordBytes.AsSpan(), safeSpan, 42, HashAlgorithmName.SHA512, 42)
    End Sub
End Class
